File: speechify-assessment-prep/src/py/tokenizer.py
"""SSML tokenizer scaffold—flesh out during practice by typing logic from memory."""
from __future__ import annotations

from typing import Dict, List, Tuple, Union

Token = Tuple[str, Union[str, Dict[str, str]], Union[Dict[str, str], None]]


def tokenize(s: str) -> List[Tuple]:
    """Convert raw SSML into token tuples (OPEN, CLOSE, SELF, TEXT).

    The heavy lifting (tag/attribute parsing) is left for assessment practice.
    """
    i, n = 0, len(s)
    out: List[Tuple] = []
    while i < n:
        if s[i] == "<":
            # TODO: determine open vs close vs self-close, parse tag + attrs.
            # Placeholder increment so the loop advances during dry runs.
            i += 1
        else:
            j = s.find("<", i)
            if j == -1:
                j = n
            text = s[i:j]
            if text:
                out.append(("TEXT", text))
            i = j
    return out

File: speechify-assessment-prep/src/ts/tokenizer.ts
// SSML tokenizer scaffold — fill in logic during practice sessions.
export type Token =
  | { kind: "OPEN"; tag: string; attrs: Record<string, string> }
  | { kind: "SELF"; tag: string; attrs: Record<string, string> }
  | { kind: "CLOSE"; tag: string }
  | { kind: "TEXT"; text: string };

export function tokenize(input: string): Token[] {
  let i = 0;
  const { length } = input;
  const out: Token[] = [];

  while (i < length) {
    if (input[i] === "<") {
      // TODO: classify tag type, parse tag name + attrs, advance index.
      // Placeholder so loop progresses during dry runs.
      i += 1;
    } else {
      const next = input.indexOf("<", i);
      const end = next === -1 ? length : next;
      const text = input.slice(i, end);
      if (text) out.push({ kind: "TEXT", text });
      i = end;
    }
  }

  return out;
}
